cluster-autoscaler/utils/drain/drain_test.go

```diff
@@ -640,75 +642,36 @@ func TestDrain(t *testing.T) {
 			expectBlockingPod:   &BlockingPod{Pod: kubeSystemRcPod, Reason: UnmovableKubeSystemPod},
 			expectDaemonSetPods: []*apiv1.Pod{},
 		},
+		{
+			description:         "Custom-controller-managed blocking pod",
+			pods:                []*apiv1.Pod{customControllerPod},
+			pdbs:                []*policyv1.PodDisruptionBudget{},
+			expectFatal:         true,
+			expectPods:          []*apiv1.Pod{},
+			expectBlockingPod:   &BlockingPod{Pod: customControllerPod, Reason: NotReplicated},
+			expectDaemonSetPods: []*apiv1.Pod{},
+		},
 	}
 
-	// run all tests for scaleDownNodesWithCustomControllerPods=false
-	// TODO(vadasambar): remove this when we get rid of scaleDownNodesWithCustomControllerPods
-	for _, test := range append(tests, testOpts{
-		// Expect blocking
-		description:         "Custom-controller-managed pod",
-		pods:                []*apiv1.Pod{customControllerPod},
-		pdbs:                []*policyv1.PodDisruptionBudget{},
-		expectFatal:         true,
-		expectPods:          []*apiv1.Pod{},
-		expectBlockingPod:   &BlockingPod{Pod: customControllerPod, Reason: NotReplicated},
-		expectDaemonSetPods: []*apiv1.Pod{},
-	}) {
-		var err error
-		var rcLister v1lister.ReplicationControllerLister
-		if len(test.rcs) > 0 {
-			rcLister, err = kube_util.NewTestReplicationControllerLister(test.rcs)
-			assert.NoError(t, err)
-		}
-		var rsLister v1appslister.ReplicaSetLister
-		if len(test.replicaSets) > 0 {
-			rsLister, err = kube_util.NewTestReplicaSetLister(test.replicaSets)
-			assert.NoError(t, err)
-		}
-
-		dsLister, err := kube_util.NewTestDaemonSetLister([]*appsv1.DaemonSet{&ds})
-		assert.NoError(t, err)
-		jobLister, err := kube_util.NewTestJobLister([]*batchv1.Job{&job})
-		assert.NoError(t, err)
-		ssLister, err := kube_util.NewTestStatefulSetLister([]*appsv1.StatefulSet{&statefulset})
-		assert.NoError(t, err)
-
-		registry := kube_util.NewListerRegistry(nil, nil, nil, nil, nil, dsLister, rcLister, jobLister, rsLister, ssLister)
-
-		pods, daemonSetPods, blockingPod, err := GetPodsForDeletionOnNodeDrain(test.pods, test.pdbs, true, true, false, registry, 0, testTime)
-
-		if test.expectFatal {
-			assert.Equal(t, test.expectBlockingPod, blockingPod)
-			if err == nil {
-				t.Fatalf("%s: unexpected non-error", test.description)
-			}
-		}
-
-		if !test.expectFatal {
-			assert.Nil(t, blockingPod)
-			if err != nil {
-				t.Fatalf("%s: error occurred: %v", test.description, err)
-			}
-		}
-
-		if len(pods) != len(test.expectPods) {
-			t.Fatalf("Wrong pod list content: %v", test.description)
-		}
+	for i := range tests {
```

Comment:
My suggestion was to duplicate the test cases and flip the flag on the copy to ensure the new flag doesn't affect the vast majority of test cases.

Reply:
I think there's another problem with the code where the test cases for flag=false are not there for shared test cases. 

> My suggestion was to duplicate the test cases and flip the flag on the copy to ensure the new flag doesn't affect the vast majority of test cases.

I see. Just to confirm my understanding. You mean something like this right?
```go
tests := []{
		// *flag: true test cases*
		{
			// shared test case 1
                       flag: true
		},
		{
			// shared test case 2
                       flag: true
		},
		{
			// additional test case 1
                       flag: true
		},
		// *flag: false test cases*
		{
			// shared test case 1 
                       flag: false
		},
		{
			// shared test case 2
                       flag: false
		},
		{
			// additional test case 1
                       flag: false
		},

}

for _, test := range tests {
	// execute test case
}
```
`shared test case 1`, `shared test case 2` and `additional test case 1` are duplicated with some minor tweaks (if required) and flipped flags.

Reply:
I think we might need a more future-proof solution here but it would also increase the code complexity.

Reply:
Maybe I can duplicate the test cases for now and there can be another issue to work through a better solution for test cases duplication/maintainability problem.

Reply:
> we would have to change name of the duplicated test cases 

This can be solved by adding the flag in the error log output [here](https://github.com/kubernetes/autoscaler/blob/2f10f76b3f4a974f5285415262ddfaf0d44cfe02/cluster-autoscaler/utils/drain/drain_test.go#L952-L968) so that the error output is something like this:
```
drain_test.go:955: Custom-controller-managed non-blocking pod: unexpected non-error, skipNodesWithCustomControllerPods: false
```

Reply:
Not sure if this is what you wanted to tell me. This can be a possible solution:
```go
sharedTests := []testOpts{
		{
			// shared test case 1
		},
		{
			// shared test case 2
		},

}

allTests = []testOpts{}
for _, sharedTest := range sharedTests {
	sharedTest.skipNodesWithCustomControllerPods = true
	allTests = append(allTests, sharedTest)
	sharedTest.skipNodesWithCustomControllerPods = false
	allTests = append(allTests, sharedTest)
}

allTests = append(allTests, testOpts{
	// additional test case 1
			   flag: true
	})

allTests = append(allTests, testOpts{
	// additional test case 1
			   flag: false
	})


for _, sharedTest := range sharedTests {
	sharedTest.newFlagInFuture = true
	allTests = append(allTests, sharedTest)
	sharedTest.newFlagInFuture = false
	allTests = append(allTests, sharedTest)
}

allTests = append(allTests, testOpts{
	// additional test case 2 for new flag in future
			   flag: false
	})

allTests = append(allTests, testOpts{
	// additional test case 2 for new flag in future
				flag: true
	})


for _, test := range tests {
	// execute test case
}

```

Reply:
I have updated this PR with a reference implementation based on the above idea. Happy to change it based on your review comments.

Reply:
LGTM, thanks!